### PR TITLE
fix(wal): release WAL read lock on rollback to prevent checkpoint busy

### LINKED ISSUE #5176: WAL Read Lock Leak After ROLLBACK Prevents Checkpoint
## Description
`ROLLBACK` of a write transaction leaks a WAL read lock, causing checkpoint to fail on shutdown. No data loss occurs.

## Reproducer
```sql
CREATE TABLE t(x);
BEGIN;
INSERT INTO t VALUES(1);
ROLLBACK;
-- On shutdown:
```

## Error Message
```
ERROR: Wal Checkpoint failed: Database is busy (repeated 4 times)
WARN: Failed to checkpoint WAL on shutdown after 3 attempts, giving up
```

`core/storage/wal.rs:2247-2252` - `reset_internal_states()` clears `max_frame_read_lock_index` to `NO_LOCK_HELD` without releasing the shared read lock. Called from `rollback()`, this leaks the read lock. Subsequent `end_read_tx()` finds `NO_LOCK_HELD` and skips unlock.

---
*This issue brought to you by Mikaël and Claude Code.*

### PR DESCRIPTION
**Pull request by aalhour**
## Description

Fix WAL read lock leak after `ROLLBACK` that caused checkpoint to fail with `Database is busy` on shutdown.

`reset_internal_states()` cleared the connection's `max_frame_read_lock_index` to `NO_LOCK_HELD` without first unlocking the corresponding `shared.read_locks[slot]`. When `pager.rollback_tx()` later called `wal.end_read_tx()`, it saw `NO_LOCK_HELD` and skipped the unlock — permanently leaking the shared read lock.

The fix has two parts:

1. **`reset_internal_states()`** now releases any held read lock before clearing the slot, encoding the safety invariant in one place.
2. **`rollback()`** now skips `reset_internal_states()` for savepoint rollbacks (`rollback_to.is_some()`), since the write transaction (and its read lock) must remain active after a savepoint rollback.

Three regression tests added:
- `test_rollback_releases_read_lock` — read lock is released after `ROLLBACK`
- `test_checkpoint_succeeds_after_rollback` — `RESTART` checkpoint succeeds (not `Busy`) after rollback
- `test_savepoint_rollback_preserves_read_lock` — read/write locks remain held after a constraint-violation savepoint rollback, and the transaction continues normally

## Motivation and context

Fixes: #5176

## Description of AI Usage

Explore call sites to make sure no edge cases weren't covered.

### CODE REVIEW

**Comment by jussisaurio** [comment]
Thanks! This looks sensible but since it's a WAL locking path I'll take some time to verify it's correct (although relevant tests are passing, ofc)

**Comment by jussisaurio** [comment]
Hi @aalhour , I wanted to add some more tests to your branch but you don't allow pushing to your fork. Could you apply this?

```diff

diff --git a/core/storage/wal.rs b/core/storage/wal.rs
index 9910c5785..55cf71add 100644
--- a/core/storage/wal.rs
+++ b/core/storage/wal.rs
@@ -3164,6 +3164,20 @@ pub mod test {
             .unwrap()
     }
 
+    #[cfg(test)]
+    fn read_slots_with_readers(shared: &WalFileShared) -> Vec<usize> {
+        shared
+            .read_locks
+            .iter()
+            .enumerate()
+            .filter_map(|(slot, lock)| {
+                let state = lock.0.load(Ordering::Acquire);
+                let has_readers = (state & super::TursoRwLock::READER_COUNT_MASK) != 0;
+                has_readers.then_some(slot)
+            })
+            .collect()
+    }
+
     fn wal_header_snapshot(shared: &Arc<RwLock<WalFileShared>>) -> (u32, u32, u32, u32) {
         // (checkpoint_seq, salt1, salt2, page_size)
         let shared_guard = shared.read();
@@ -4085,6 +4099,80 @@ pub mod test {
         }
     }
 
+    #[test]
+    fn test_rollback_releases_shared_read_lock_slot() {
+        let (db, _path) = get_database();
+        let conn = db.connect().unwrap();
+
+        conn.execute("CREATE TABLE t(x)").unwrap();
+        conn.execute("BEGIN").unwrap();
+        conn.execute("INSERT INTO t VALUES(1)").unwrap();
+
+        let locked_slots_before = {
+            let shared = db.shared_wal.read();
+            read_slots_with_readers(&shared)
+        };
+        assert_eq!(
+            locked_slots_before.len(),
+            1,
+            "expected exactly one shared read-lock slot while transaction is active"
+        );
+
+        conn.execute("ROLLBACK").unwrap();
+
+        let locked_slots_after = {
+            let shared = db.shared_wal.read();
+            read_slots_with_readers(&shared)
+        };
+        assert!(
+            locked_slots_after.is_empty(),
+            "ROLLBACK must release the shared read-lock slot"
+        );
+    }
+
+    #[test]
+    fn test_rollback_releases_slot_zero_read_lock() {
+        let (db, _path) = get_database();
+        let conn = db.connect().unwrap();
+
+        conn.execute("CREATE TABLE test(id integer primary key, value text)")
+            .unwrap();
+        bulk_inserts(&conn, 3, 3);
+        {
+            let pager = conn.pager.load();
+            let result = run_checkpoint_until_done(&pager, CheckpointMode::Restart);
+            assert!(
+                result.everything_backfilled(),
+                "restart checkpoint setup must fully backfill WAL"
+            );
+        }
+
+        conn.execute("BEGIN").unwrap();
+        conn.execute("INSERT INTO test(value) VALUES('slot0')")
+            .unwrap();
+
+        let locked_slots_before = {
+            let shared = db.shared_wal.read();
+            read_slots_with_readers(&shared)
+        };
+        assert_eq!(
+            locked_slots_before,
+            vec![0],
+            "writer should use slot 0 when WAL is fully checkpointed"
+        );
+
+        conn.execute("ROLLBACK").unwrap();
+
+        let locked_slots_after = {
+            let shared = db.shared_wal.read();
+            read_slots_with_readers(&shared)
+        };
+        assert!(
+            locked_slots_after.is_empty(),
+            "ROLLBACK must release slot 0 shared read-lock as well"
+        );
+    }
+
     #[test]
     fn test_savepoint_rollback_preserves_read_lock() {
         let (db, _path) = get_database();
@@ -4129,6 +4217,77 @@ pub mod test {
         assert_eq!(count, 1, "first insert should survive savepoint rollback");
     }
 
+    #[test]
+    fn test_savepoint_then_tx_rollback_allows_restart_checkpoint_from_other_connection() {
+        let (db, _path) = get_database();
+        let conn1 = db.connect().unwrap();
+        let conn2 = db.connect().unwrap();
+
+        conn1
+            .execute("CREATE TABLE test(id integer primary key, value text)")
+            .unwrap();
+        bulk_inserts(&conn1, 2, 2);
+        let count_before = count_test_table(&conn1);
+
+        conn1.execute("BEGIN").unwrap();
+        conn1
+            .execute("INSERT INTO test(id, value) VALUES(1000, 'first')")
+            .unwrap();
+        let duplicate = conn1.execute("INSERT INTO test(id, value) VALUES(1000, 'dup')");
+        assert!(duplicate.is_err(), "duplicate PK insert must fail");
+
+        {
+            let pager = conn1.pager.load();
+            let wal = pager.wal.as_ref().unwrap();
+            assert!(
+                wal.holds_read_lock(),
+                "read lock must still be held after savepoint rollback"
+            );
+            assert!(
+                wal.holds_write_lock(),
+                "write lock must still be held after savepoint rollback"
+            );
+        }
+
+        conn1.execute("ROLLBACK").unwrap();
+
+        {
+            let pager = conn1.pager.load();
+            let wal = pager.wal.as_ref().unwrap();
+            assert!(
+                !wal.holds_read_lock(),
+                "read lock must be released after transaction rollback"
+            );
+            assert!(
+                !wal.holds_write_lock(),
+                "write lock must be released after transaction rollback"
+            );
+        }
+
+        let locked_slots_after_rollback = {
+            let shared = db.shared_wal.read();
+            read_slots_with_readers(&shared)
+        };
+        assert!(
+            locked_slots_after_rollback.is_empty(),
+            "transaction rollback after savepoint failure must not leak shared read locks"
+        );
+        assert_eq!(
+            count_test_table(&conn1),
+            count_before,
+            "transaction rollback should remove writes made before savepoint failure"
+        );
+
+        let result = {
+            let pager = conn2.pager.load();
+            run_checkpoint_until_done(&pager, CheckpointMode::Restart)
+        };
+        assert!(
+            result.everything_backfilled(),
+            "restart checkpoint from another connection must succeed after full rollback"
+        );
+    }
+
     #[test]
     fn test_checkpoint_succeeds_after_rollback() {
         let (db, _path) = get_database();


```

**Comment by aalhour** [comment]
Hi @jussisaurio, thanks for taking a look, I just added your tests to the branch, let me know what you think 🙏

**Comment by jussisaurio** [comment]
@aalhour I am legitimately not sure about the browser bindings test failure here - i reran the CI and it happened again, so might be a persistent error now. Needs some investigation

**Comment by aalhour** [comment]
@jussisaurio, I am investigating the issue, it looks persistent, I will push more changes to this branch, then we'll take a second look and see where to take it next.

**Comment by aalhour** [comment]
Hi @jussisaurio, I tracked down the CI failure in `promise.test.ts` (`on-disk db large inserts`) and confirmed it was a combination of two issues:

1. Checkpoint could run while the statement still held an implicit read tx (`RESTART/TRUNCATE` self-conflict path).
2. In browser WASM truncate checkpoint flow, we called sync `db_file.size()`, which is worker-only in OPFS bindings.  
   That triggered `size can be called only from web worker context`, and the later `BorrowMutError` was a follow-on panic.

### Fix in this PR
- `core/vdbe/execute.rs`
  - In `op_checkpoint`, for `RESTART | TRUNCATE`, end implicit read tx first (`pager.end_read_tx()`) when held.
- `core/storage/pager.rs`
  - In `CheckpointPhase::TruncateDbFile`:
    - keep native optimization (`expected >= self.db_file.size()?`)
    - WASM-only: skip sync `size()` pre-check and continue via existing `truncate(...)` path.
- `tests/integration/wal/test_wal.rs`
  - Added regression: `test_truncate_checkpoint_not_busy_after_rollback`.

### Verification

- Native:
  - `cargo test -p core_tester --test integration_tests wal::test_wal::test_truncate_checkpoint_not_busy_after_rollback -- --nocapture`
- Browser CI-analog (Docker, rebuilt wasm from patched Rust):
  - `CI=1 npm exec vitest -- --browser=chromium --run promise.test.ts -t "on-disk db large inserts"`

If reviewers want, I can also add a browser-focused regression that asserts this checkpoint path explicitly.

**Review by PThorpe92 — APPROVED** [review]
Looks good 👍
